Define new ILGen optimization to induce OSR for unresolved value types

hzongaro opened the pull request:
Define an OpenJ9-specific optimization, `HandleRecompilationOps`, that looks for `ResolveCHK` operations that are applied to fields of value type or that are applied to a `loadaddr` on a value type class that is being applied to a `newvalue` operation.  As the JIT is not able to generate code that performs the required operations on such unresolved types, it
instead will attempt to induce a recompilation request, triggering OSR.  If that's not possible, it will abort the compilation.

This pull request depends upon [OpenJ9 pull request #11020](https://github.com/eclipse/openj9/pull/11020) and [OMR pull request #5631](https://github.com/eclipse/omr/pull/5631).

Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

hzongaro committed:
Define new ILGen optimization to induce OSR for unresolved value types

Define an OpenJ9-specific optimization, handleRecompilationOps, that
looks for ResolveCHK operations that are applied to fields of value type
or that are applied to a loadaddr on a value type class that is being
applied to a newvalue operation.  As the JIT is not able to generate
code that performs the required operations on such unresolved types, it
instead will attempt to induce a recompilation request, triggering OSR.
If that's not possible, it will abort the compilation.

Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

andrewcraik commented on runtime/compiler/optimizer/HandleRecompilationOps.cpp:
```diff
@@ -0,0 +1,258 @@
+int32_t
+TR_HandleRecompilationOps::perform()
+   {
+   _enableTransform = true;
```
Why not just return 0 from the cases checked when we know we don't want to do this rather than having the global variable?

hzongaro replied:
The way the changes are set up right now, `Walker.cpp` is always generating IL for these situations rather than aborting the compilation.  It's now relying on this ILGen opt. pass to check whether inducing OSR is needed, and this opt. aborts if it cannot induce OSR.

If you feel that's the wrong approach, and that `Walker.cpp` itself should abort the compilation if OSR can't be induced, I'll switch to that way of handling it.

hzongaro replied:
Here's what I'm thinking of doing.  The conditions that are checked in `TR_HandleRecompilationOps::perform()` will also be checked during ILGen in `Walker.cpp`.  If any of those conditions is not satisfied, compilation will be aborted during ILGen itself.  That way, if `TR_HandleRecompilationOps::perform` encounters any of those conditions, it can immediately return without performing any analysis.

Some additional conditions are checked in [`OMR::ResolvedMethodSymbol::induceOSRAfterAndRecompile`](https://github.com/hzongaro/omr/blob/e2fac34fcb3571cdf5351693a04fc2852b18f46f/compiler/il/OMRResolvedMethodSymbol.cpp#L478), which is called by [`TR_HandleRecompilationOps::visitNode`](https://github.com/eclipse/openj9/blob/5a773894edd1f76514c741eb425971d9c6cd3065/runtime/compiler/optimizer/HandleRecompilationOps.cpp#L219).  The result of that call to `OMR::ResolvedMethodSymbol::induceOSRAfterAndRecompile` is currently checked with a `TR_ASSERT_FATAL`, but if the conditions that `induceOSRAfterAndRecompile` is checking could not be checked during ILGen itself, then `TR_HandleRecompilationOps::visitNode` should simply abort the compilation at that point rather than performing a fatal assertion. . . .

hzongaro replied:
Made these changes in commit 3ecec562fd3b56fad1eeef986b6ca9d71d4aee17

andrewcraik commented on runtime/compiler/optimizer/HandleRecompilationOps.cpp:
```diff
@@ -0,0 +1,258 @@
+              nextTree = nextTree->getNextTreeTop())
+            {
+            TR::Node *nextNode = nextTree->getNode();
+            if (nextNode->getOpCode().isTreeTop() && nextNode->getNumChildren() > 0)
```
you may need || isCheck or similar...

hzongaro replied:
Thanks!  I'll fix that.

hzongaro replied:
Fixed in commit 3ecec562fd3b56fad1eeef986b6ca9d71d4aee17

andrewcraik commented on runtime/compiler/optimizer/HandleRecompilationOps.cpp:
```diff
@@ -0,0 +1,258 @@
+
+      if (loadAddrSymRef->isUnresolved())
+         {
+         for (TR::TreeTop *nextTree = currTree->getNextTreeTop();
```
You might want to look at the TreeTop iterator in the walkers to save some boilerplate

hzongaro replied:
I neglected to address this comment.  I'll take care of this next.

hzongaro replied:
Addressed in commit 305c99e431ebe5ac8995080f366479e4df9e43b6

hzongaro commented on runtime/compiler/optimizer/HandleRecompilationOps.cpp:
```diff
@@ -0,0 +1,258 @@
+     traceMsg(comp(), "Exiting HandleRecompilationOps\n");
+     }
+
+   return 0;
```
Oops!  Thanks for catching that.  I'll fix it.

hzongaro replied:
Fixed in commit 3ecec562fd3b56fad1eeef986b6ca9d71d4aee17

hzongaro committed:
Address code review comments

Reworked to have Walker.cpp abort compilation in most cases if OSR
cannot be induced with recompilation requested rather than always
delaying aborting until HandleRecompilationOps is run.  The conditions
that are checked have been moved into a common method,
TR::Compilation::isOSRAllowedForOperationsRequiringRecompilation.
That allows HandleRecompilationOps::perform to exit early in cases where
the tranformation it performs is absolutely not needed.

Also corrected HandleRecompilationOps::perform to return 1 if it
performs any transformation.

Finally, changed signature of visitNode to accept only a tree to make
it clear that it wouldn't be recursively called for all the nodes in
the tree, and renamed it to visitTree.

Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

hzongaro commented on runtime/compiler/optimizer/HandleRecompilationOps.cpp:
```diff
@@ -0,0 +1,258 @@
+            cleanupTT = nextTT;
+            }
+
+         TR_ASSERT_FATAL(_methodSymbol->induceOSRAfterAndRecompile(currTree, node->getByteCodeInfo(), branchTT, false, 0, &lastTT), "Unable to generate induce OSR");
```
Corrected this in commit 3ecec562fd3b56fad1eeef986b6ca9d71d4aee17

andrewcraik commented on runtime/compiler/optimizer/HandleRecompilationOps.cpp:
```diff
@@ -0,0 +1,242 @@
+   }
+
+bool
+TR_HandleRecompilationOps::visitNode(TR::TreeTop *currTree, TR::Node *node)
```
given that this is not recursive it might be cleaner just to pass the tree and pull out the node otherwise someone might think this might recurse...

hzongaro replied:
Quite right!  At first I was thinking it would be recursive, but then I never cleaned up the signature/name.  Fixed in commit bf856a82dd87a1eed2fc0dd8001d7df9b818b72d

vijaysun-omr commented:
For OSR changes, testing is particularly important since the code invariably gets into detailed bookkeeping to make the OSR transition occur into the interpreter. How will these changes be tested ? Has there been some limited special mode testing done already, e.g. at count=0 to verify that the new OSR transitions are working correctly for unresolved value types ?

hzongaro commented:
Vijay @vijaysun-omr asked,

> How will these changes be tested? Has there been some limited special mode testing done already

Yes, I've written some value types unit tests that were recently merged from commit fbba1911401fd4be37d50554604842503481cb42.  These are run with `count=1,disableAsyncCompilation`, among other variations.  I ran with and without `TR_EnableOSRAtAllOptLevels=yes` as the methods involved were being compiled at `warm` opt level.

The tests themselves run methods several times that contain `defaultvalue`, `withfield`, `putfield` and `getfield` operations on value types or fields of value types.  The operations are guarded so that the method might be compiled while the types involved are still unresolved.  After several calls, the condition that guards those operations is made to be satisfied so that the operation is attempted, inducing the OSR transition.

I examined the logs produced to verify that the methods were compiled and subsequently recompiled as expected.

This is admittedly a very small sample size.

hzongaro committed:
Remove unused edges from CFG

Upon inserting an unconditional branch to an OSR induce block at the end
of a block, any edges to existing successors should be removed as they
are no longer relevant or correct.

Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

hzongaro commented:
> >    How will these changes be tested? Has there been some limited special mode testing done already

> Yes, I've written some value types unit tests that were recently merged from commit fbba191. These are run with `count=1,disableAsyncCompilation`, among other variations. I ran with and without `TR_EnableOSRAtAllOptLevels=yes` as the methods involved were being compiled at warm opt level.

I did some further runs with `count=0`, and found a problem with the state of the CFG after this optimization is applied.  That's fixed in commit 73b5212a941666f0d8c8406c0536d6bd47e6c1a4.  I've also rebased the changes onto the latest version of the code.

I believe this is now ready for review and am marking it as such.

hzongaro committed:
Check if HandleRecompilationOps is enabled for unresolved value types

J9::Compilation::isOSRAllowedForOperationsRequiringRecompilation
checks various conditions that must be satisfied in order to induce
OSR and request recompilation for an operation on a value type that is
unresolved.  That the HandleRecompilationOps is enabled must be
included, as it is responsible for recognizing such operations and
generating trees to induce OSR.

Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

hzongaro commented:
I just tripped across another bug, so I've moved this back to draft status.  I will fix that problem, do some more runs and testing to see whether it shakes out any other problems, and then I will move this back to "ready for review."

Sorry for the churn!

hzongaro commented:
Marking as "Ready for review" again.

I've left a couple of small commits in this pull request, just to make it clear what the recent changes have been.  I will squash them down to a reasonable set once all reviews have been completed.

vijaysun-omr commented:
Good to hear that you have a targeted testing plan and used it to shake out more bugs since the last review. 

Taking a step back, you mentioned in an early comment in this PR that this optimization "looks for ResolveCHK operations that are applied to fields of value type or that are applied to a loadaddr on a value type class that is being applied to a newvalue operation." Do you know in every case that a particular unresolved type is in fact an unresolved value type ? i.e. does the fact that other code (e.g. the type of the base object used to do the unresolved field access) could be unresolved not make detecting when one is dealing with an unresolved value type a challenge ?

vijaysun-omr commented:
The other high level question is for you is to mention an example where not using OSR in these situations could cause complications to the JIT compiler. I'm trying to make sure that using OSR in this situation has enough up side to make up for the down sides of the complication of using OSR in several places.

hzongaro commented:
Vijay @vijaysun-omr, thank you for the questions and comments.

> Taking a step back, you mentioned in an early comment in this PR that this optimization "looks for ResolveCHK operations that are applied to fields of value type or that are applied to a loadaddr on a value type class that is being applied to a newvalue operation." Do you know in every case that a particular unresolved type is in fact an unresolved value type ? i.e. does the fact that other code (e.g. the type of the base object used to do the unresolved field access) could be unresolved not make detecting when one is dealing with an unresolved value type field a challenge ?

In the case of a field whose type is a value type, the signature of the field will have the letter `Q` as a prefix rather than `L`, so that will signal that the field is a value type.  Just in case my wording was ambiguous or at least unclear, by "fields of value type" I meant, "fields whose types are value types" rather than "the fields in a value type class."

In the case of an unresolved class whose address is being loaded for a `newvalue` operation, if the class was not a value type, an `IncompatibleClassChangeError` will result.  Triggering OSR in that circumstance has a built-in assumption that that won't be a situation requiring peak performance an `IncompatibleClassChangeError` .

hzongaro commented:
> The other high level question is for you is to mention an example where not using OSR in these situations could cause complications to the JIT compiler. I'm trying to make sure that using OSR in this situation has enough up side to make up for the down sides of the complication of using OSR in several places.

In the case of a `newvalue` operation, the JIT generates IL during lowering that explicitly sets the fields of the new instance of the value type.  If the class is unresolved, the JIT doesn't know the fields of the type, and is unable to generate IL to set the fields.

In the case of `putfield` or `getfield` where the field's type is a value type, that field's fields could be flattened into the class that contains that field.  Generating IL to access the field depends on knowing whether its fields are flattened, which depends on its type having been resolved.

Currently the JIT compiler does not have a mechanism for handling these operations on unresolved value types, so it aborts the compilation. 

> Another perspective along similar lines to my last couple of comments is to think carefully if there are cases when this OSR based approach would not be an option and if there are any such, then maybe you need a non-OSR based approach as well in addition to this one, and at that point, one could ask if we just go with the non-OSR solution or not.

I'm not aware of cases that the OSR approach won't handle when OSR with recompilation is available, but clearly there are cases it doesn't handle when OSR is not available.

It might be possible to define VM helpers that could handle the unresolved cases.  That would allow the JIT to avoid aborting compilation in the non-OSR case, and then we could evaluate the benefit of the OSR approach to handling the unresolved cases versus using helpers.

I would need to explore further whether defining such helpers would be feasible.

Leonardo2718 commented:
@vijaysun-omr 

> The other high level question is for you is to mention an example where not using OSR in these situations could cause complications to the JIT compiler. I'm trying to make sure that using OSR in this situation has enough up side to make up for the down sides of the complication of using OSR in several places.
> 
> Another perspective along similar lines to my last couple of comments is to think carefully if there are cases when this OSR based approach would not be an option and if there are any such, then maybe you need a non-OSR based approach as well in addition to this one, and at that point, one could ask if we just go with the non-OSR solution or not. I'm not suggesting any change in design but just asking questions to get the rationale straight for me and documented in the open.

When the JIT encounters a field whose type is unresolved, the fundamental issue we run into is that we don't know the layout of the type or if any special handling is required. Without knowing the layout of the type, the JIT can either abort compilation (which is what we currently do) or generate code that can handle any arbitrary layout and any case that requires special handling (e.g. an abstract base class that may have value type subclasses). Such native code would have to duplicate much of the functionality of the interpreter, which is impractical. VM helpers would be impractical to implement for similar reasons and would essentially reduce the JIT to a template compiler. So, we think that the best option to avoid aborting compilation is to transition execution to the interpreter (which already knows how to handle all possible cases) using OSR. My understanding is that we expect compiled code to hit unresolved fields petty infrequently, which is the kind of scenario OSR is good for.

A possible alternative could be to use an approach based on inline caching. I've considered implementing certain (resolved) value type operations using PICs and this could be an extension/generalization of that idea. Essentially, the JIT could generate inline caches for operations on unresolved fields. Resolution of the field would then trigger the transition of the cache to an appropriate monomorphic state based not the type of the field. However, I believe such a mechanism would be quite complex and time-consuming to implement. I don't think any of the existing PIC infrastructure is in a state where it could be reused for this purpose either. So, given that the immediate goal is to avoid aborting compilation on unresolved fields, I believe OSR to be the more practical option.

I hope this clarifies things :slightly_smiling_face:

vijaysun-omr commented:
Thanks @hzongaro and @Leonardo2718 for the clarifications. They definitely helped. 

"VM helpers would be impractical to implement for similar reasons and would essentially reduce the JIT to a template compiler." is probably true for only the unresolved code paths that are expected to be rare. 
I'm unsure how many VM helpers we would be adding and what would be the logic in them to assess how much of the interpreter functionality would be duplicated in these VM helpers. Note that we do have VM helpers for several IL opcodes that could be considered a duplication of interpreter functionality by using similar reasoning, (e.g. the VM helper to throw a NPE or AIOB exception) where in theory we could have used OSR as well as an alternative. I think the level of duplication (i.e. the number of VM helpers we might need) would help me assess how clear the case is for just using OSR. I'd also appreciate @0xdaryl opinion on this. 

I do agree with Leo that a PIC based solution is probably less ideal than even an OSR based one. I don't know if we have a strong case for using PIC for unresolved code in general given how unlikely they are to execute at the high count values we use to do JIT compilations. It may be that just calling out to VM helpers every time for unresolved may well be performant and won't incur as much of a maintenance cost in general, but as I said, I'm certainly in agreement with not going down that path for this design.

Leonardo2718 commented:
> I'm unsure how many VM helpers we would be adding and what would be the logic in them to assess how much of the interpreter functionality would be duplicated in these VM helpers. Note that we do have VM helpers for several IL opcodes that could be considered a duplication of interpreter functionality by using similar reasoning, (e.g. the VM helper to throw a NPE or AIOB exception) where in theory we could have used OSR as well as an alternative. I think the level of duplication (i.e. the number/complexity of VM helpers we might need) would help me assess how clear the case is for just using OSR.

We actually do use VM helpers pretty heavily for the resolved paths. The plan is for the helpers to eventually become the slow paths for these operations while we generate native code for the fast paths. I don't have a good sense of how much work it would be to change the helpers to support the unresolved paths, maybe @tajila can comment? I believe the helpers that would have to be changed are for `getfield`, `withfield`, and `putfield`. We will probably also need a new helper for `defaultvalue`. I do know the interfaces of these helpers would have to change, so the JIT code to generate the dispatch sequences would also have to change.

With that said, if we're able to use helpers for unresolved cases, then I believe the only benefit OSR would provide is the ability to trigger a recompilation of the method after resolution. I don't know that we currently have a mechanism for triggering recompilation from a helper call.

hzongaro commented:
> I believe the helpers that would have to be changed are for `getfield`, `withfield`, and `putfield`. We will probably also need a new helper for `defaultvalue`.

Or a parallel set of very similar helpers, if that would make more sense, with the existing helpers used in the resolved, flattened case, and new helpers for the unresolved case.

tajila commented:
> > I believe the helpers that would have to be changed are for getfield, withfield, and putfield. We will probably also need a new helper for defaultvalue.

> Or a parallel set of very similar helpers, if that would make more sense, with the existing helpers used in the resolved, flattened case, and new helpers for the unresolved case.

For some of the existing helpers we are at the limit for the number of args we can pass by register. We would need to pass the j9method in addition to the existing args to perform a resolution. So, in order to reuse them we would need to either start passing some args in static space or create a new helper just for resolution (preferred choice).

hzongaro commented:
I've moved this pull request back to draft status, while we explore the possibility of using helpers instead of OSR to handle unresolved fields or classes.

vijaysun-omr commented:
Thanks, another thing to think about in this overall design is the compile time cost of running OSR at cold optimization level (that I believe is not done currently) to handle unresolved value types. During startup typically, there are at least a few cold JIT compilations done (even when SCC/AOT is in use) and slowing those down substantially may have a noticeable impact on startup time overall. Note that cold optimization level pretty much does no dataflow analyses whereas OSR could bring some dataflow analyses that are usually more expensive than local analyses. FYI @mpirvu and @dsouzai

genie-openj9 commented:
Can one of the admins verify this patch?

hzongaro commented:
The notion of using OSR to deal with unresolved fields with value types enabled has been abandoned.  Closing this pull request.
